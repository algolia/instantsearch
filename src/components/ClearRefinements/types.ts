import { ClearRefinementsRendererOptions } from '../../connectors/clear-refinements/connectClearRefinements';
import {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';

export type ClearRefinementsProps = {
  refine: ClearRefinementsRendererOptions['refine'];
  cssClasses: ClearRefinementsCSSClasses;
  hasRefinements: ClearRefinementsRendererOptions['hasRefinements'];
  templateProps: {
    [key: string]: any;
    templates: ClearRefinementsTemplates;
  };
};
