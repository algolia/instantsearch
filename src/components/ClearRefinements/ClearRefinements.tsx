/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';
import Template from '../Template/Template';
import { ClearRefinementsRenderState } from '../../connectors/clear-refinements/connectClearRefinements';
import {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';

type ClearRefinementsComponentCSSClasses = Required<
  {
    [key in keyof ClearRefinementsCSSClasses]: string;
  }
>;

export type ClearRefinementsProps = {
  refine: ClearRefinementsRenderState['refine'];
  cssClasses: ClearRefinementsComponentCSSClasses;
  hasRefinements: ClearRefinementsRenderState['hasRefinements'];
  templateProps: {
    [key: string]: any;
    templates: ClearRefinementsTemplates;
  };
};

const ClearRefinements = ({
  hasRefinements,
  refine,
  cssClasses,
  templateProps,
}: ClearRefinementsProps) => (
  <div className={cssClasses.root}>
    <Template
      {...templateProps}
      templateKey="resetLabel"
      rootTagName="button"
      rootProps={{
        className: cx(cssClasses.button, {
          [cssClasses.disabledButton]: !hasRefinements,
        }),
        onClick: refine,
        disabled: !hasRefinements,
      }}
      data={{ hasRefinements }}
    />
  </div>
);

export default ClearRefinements;
