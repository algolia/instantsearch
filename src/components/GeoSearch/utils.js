export const getBoundsZoomLevel = (bounds, mapDim) => {
  const WORLD_DIM = { height: 256, width: 256 };
  const ZOOM_MAX = 21;

  const latRad = lat => {
    const sin = Math.sin(lat * Math.PI / 180);
    const radX2 = Math.log((1 + sin) / (1 - sin)) / 2;

    return Math.max(Math.min(radX2, Math.PI), -Math.PI) / 2;
  };

  const zoom = (mapPx, worldPx, fraction) =>
    Math.floor(Math.log(mapPx / worldPx / fraction) / Math.LN2);

  const ne = bounds.getNorthEast();
  const sw = bounds.getSouthWest();

  const latFraction = (latRad(ne.lat()) - latRad(sw.lat())) / Math.PI;

  const lngDiff = ne.lng() - sw.lng();
  const lngFraction = (lngDiff < 0 ? lngDiff + 360 : lngDiff) / 360;

  const latZoom = zoom(mapDim.height, WORLD_DIM.height, latFraction);
  const lngZoom = zoom(mapDim.width, WORLD_DIM.width, lngFraction);

  return Math.min(latZoom, lngZoom, ZOOM_MAX);
};
