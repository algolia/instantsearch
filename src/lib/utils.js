import reduce from 'lodash/collection/reduce';
import forEach from 'lodash/collection/forEach';
import find from 'lodash/collection/find';
import get from 'lodash/object/get';
import isEmpty from 'lodash/lang/isEmpty';
import keys from 'lodash/object/keys';
import uniq from 'lodash/array/uniq';
import mapKeys from 'lodash/object/mapKeys';

export {
  getContainerNode,
  bemHelper,
  prepareTemplateProps,
  isSpecialClick,
  isDomElement,
  getRefinements,
  clearRefinementsFromState,
  clearRefinementsAndSearch,
  prefixKeys
};

/**
 * Return the container. If it's a string, it is considered a
 * css selector and retrieves the first matching element. Otherwise
 * test if it validates that it's a correct DOMElement.
 * @param {string|DOMElement} selectorOrHTMLElement a selector or a node
 * @return {DOMElement} The resolved DOMElement
 * @throws Error when the type is not correct
 */
function getContainerNode(selectorOrHTMLElement) {
  const isFromString = typeof selectorOrHTMLElement === 'string';
  let domElement;
  if (isFromString) {
    domElement = document.querySelector(selectorOrHTMLElement);
  } else {
    domElement = selectorOrHTMLElement;
  }

  if (!isDomElement(domElement)) {
    let errorMessage = 'Container must be `string` or `HTMLElement`.';
    if (isFromString) {
      errorMessage += ` Unable to find ${selectorOrHTMLElement}`;
    }
    throw new Error(errorMessage);
  }

  return domElement;
}

/**
 * Returns true if the parameter is a DOMElement.
 * @param {any} o the value to test
 * @return {boolean} true if o is a DOMElement
 */
function isDomElement(o) {
  return o instanceof window.HTMLElement || Boolean(o) && o.nodeType > 0;
}

function isSpecialClick(event) {
  const isMiddleClick = event.button === 1;
  return isMiddleClick || event.altKey || event.ctrlKey || event.metaKey || event.shiftKey;
}

/**
 * Creates BEM class name according the vanilla BEM style.
 * @param {string} block the main block
 * @return {function} function that takes up to 2 parameters
 * that determine the element and the modifier of the BEM class.
 */
function bemHelper(block) {
  return function(element, modifier) {
    // block--element
    if (element && !modifier) {
      return `${block}--${element}`;
    }
    // block--element__modifier
    if (element && modifier) {
      return `${block}--${element}__${modifier}`;
    }
    // block__modifier
    if (!element && modifier) {
      return `${block}__${modifier}`;
    }

    return block;
  };
}

/**
 * Prepares an object to be passed to the Template widget
 * @param {object} unknownBecauseES6 an object with the following attributes:
 *  - transformData
 *  - defaultTemplate
 *  - templates
 *  - templatesConfig
 * @return {object} the configuration with the attributes:
 *  - transformData
 *  - defaultTemplate
 *  - templates
 *  - useCustomCompileOptions
 */
function prepareTemplateProps({
  transformData,
  defaultTemplates,
  templates,
  templatesConfig
}) {
  const preparedTemplates = prepareTemplates(defaultTemplates, templates);

  return {
    transformData,
    templatesConfig,
    ...preparedTemplates
  };
}

function prepareTemplates(defaultTemplates = {}, templates = {}) {
  const allKeys = uniq([...(keys(defaultTemplates)), ...(keys(templates))]);

  return reduce(allKeys, (config, key) => {
    const defaultTemplate = defaultTemplates[key];
    const customTemplate = templates[key];
    const isCustomTemplate = customTemplate !== undefined && customTemplate !== defaultTemplate;

    config.templates[key] = isCustomTemplate ? customTemplate : defaultTemplate;
    config.useCustomCompileOptions[key] = isCustomTemplate;

    return config;
  }, {templates: {}, useCustomCompileOptions: {}});
}

function getRefinement(state, type, attributeName, name, resultsFacets) {
  const res = {type, attributeName, name};
  let facet = find(resultsFacets, {name: attributeName});
  let count;
  if (type === 'hierarchical') {
    const facetDeclaration = state.getHierarchicalFacetByName(attributeName);
    const splitted = name.split(facetDeclaration.separator);
    res.name = splitted[splitted.length - 1];
    for (let i = 0; facet !== undefined && i < splitted.length; ++i) {
      facet = find(facet.data, {name: splitted[i]});
    }
    count = get(facet, 'count');
  } else {
    count = get(facet, `data["${res.name}"]`);
  }
  const exhaustive = get(facet, 'exhaustive');
  if (count !== undefined) {
    res.count = count;
  }
  if (exhaustive !== undefined) {
    res.exhaustive = exhaustive;
  }
  return res;
}

function getRefinements(results, state) {
  const res = [];

  forEach(state.facetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'facet', attributeName, name, results.facets));
    });
  });

  forEach(state.facetsExcludes, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push({type: 'exclude', attributeName, name, exclude: true});
    });
  });

  forEach(state.disjunctiveFacetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'disjunctive', attributeName, name, results.disjunctiveFacets));
    });
  });

  forEach(state.hierarchicalFacetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'hierarchical', attributeName, name, results.hierarchicalFacets));
    });
  });

  forEach(state.numericRefinements, (operators, attributeName) => {
    forEach(operators, (values, operator) => {
      forEach(values, value => {
        res.push({
          type: 'numeric',
          attributeName,
          name: `${value}`,
          numericValue: value,
          operator
        });
      });
    });
  });

  forEach(state.tagRefinements, name => {
    res.push({type: 'tag', attributeName: '_tags', name});
  });

  return res;
}

function clearRefinementsFromState(inputState, attributeNames) {
  let state = inputState;

  if (isEmpty(attributeNames)) {
    state = state.clearTags();
    state = state.clearRefinements();
    return state;
  }

  forEach(attributeNames, attributeName => {
    if (attributeName === '_tags') {
      state = state.clearTags();
    } else {
      state = state.clearRefinements(attributeName);
    }
  });

  return state;
}

function clearRefinementsAndSearch(helper, attributeNames) {
  helper.setState(clearRefinementsFromState(helper.state, attributeNames)).search();
}

function prefixKeys(prefix, obj) {
  if (obj) {
    return mapKeys(obj, (v, k) => prefix + k);
  }

  return undefined;
}
