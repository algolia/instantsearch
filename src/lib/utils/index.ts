export { default as capitalize } from './capitalize';
export { default as defer } from './defer';
export { default as isDomElement } from './isDomElement';
export { default as getContainerNode } from './getContainerNode';
export { default as isSpecialClick } from './isSpecialClick';
export { default as prepareTemplateProps } from './prepareTemplateProps';
export { default as renderTemplate } from './renderTemplate';
export { default as getRefinements } from './getRefinements';
export { default as clearRefinements } from './clearRefinements';
export { default as escapeRefinement } from './escapeRefinement';
export { default as enhanceConfiguration } from './enhanceConfiguration';
export { default as unescapeRefinement } from './unescapeRefinement';
export { default as checkRendering } from './checkRendering';
export { default as getPropertyByPath } from './getPropertyByPath';
export { default as noop } from './noop';
export { default as isFiniteNumber } from './isFiniteNumber';
export { default as isPlainObject } from './isPlainObject';
export { default as uniq } from './uniq';
export { default as range } from './range';
export { default as isEqual } from './isEqual';
export { default as escape } from './escape';
export { default as find } from './find';
export { default as mergeDeep } from './mergeDeep';
export { default as resolveSearchParameters } from './resolveSearchParameters';
export { warning, deprecate } from './logger';
export {
  createDocumentationLink,
  createDocumentationMessageGenerator,
} from './documentation';
export {
  aroundLatLngToPosition,
  insideBoundingBoxToBoundingBox,
} from './geo-search';
export { addAbsolutePosition } from './hits-absolute-position';
export { addQueryID } from './hits-query-id';
