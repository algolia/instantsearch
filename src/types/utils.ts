/* eslint-disable-next-line @typescript-eslint/generic-type-naming */
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
