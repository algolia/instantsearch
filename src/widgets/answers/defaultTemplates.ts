import { AnswersComponentTemplates } from '../../components/Answers/Answers';

const defaultTemplates: AnswersComponentTemplates = {
  header: '',
  loader: '',
  item: (item) => JSON.stringify(item),
};

export default defaultTemplates;
