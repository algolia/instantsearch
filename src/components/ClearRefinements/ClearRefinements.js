import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';
import Template from '../Template';
import { isSpecialClick } from '../../lib/utils';

export default class RawClearRefinements extends Component {
  componentWillMount() {
    this.handleClick = this.handleClick.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.url !== nextProps.url ||
      this.props.hasRefinements !== nextProps.hasRefinements
    );
  }

  handleClick(e) {
    if (isSpecialClick(e)) {
      // do not alter the default browser behavior
      // if one special key is down
      return;
    }
    e.preventDefault();
    this.props.refine();
  }

  render() {
    const { hasRefinements, cssClasses } = this.props;
    const data = { hasRefinements };

    const className =
      !hasRefinements && cssClasses.disabledButton
        ? `${cssClasses.button} ${cssClasses.disabledButton}`
        : cssClasses.button;

    return (
      <div className={cssClasses.root}>
        <Template
          rootTagName="button"
          rootProps={{
            className,
            href: this.props.url,
            onClick: this.handleClick,
            disabled: !hasRefinements,
          }}
          data={data}
          templateKey="button"
          {...this.props.templateProps}
        />
      </div>
    );
  }
}

RawClearRefinements.propTypes = {
  refine: PropTypes.func.isRequired,
  cssClasses: PropTypes.shape({
    root: PropTypes.string,
    button: PropTypes.string,
    disabledButton: PropTypes.string,
  }),
  hasRefinements: PropTypes.bool.isRequired,
  templateProps: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
};
