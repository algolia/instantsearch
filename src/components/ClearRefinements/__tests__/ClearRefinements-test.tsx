/**
 * @jest-environment jsdom
 */
/** @jsx h */

import { h } from 'preact';
import ClearRefinements from '../ClearRefinements';
import { mount } from '../../../../test/utils/enzyme';
import { prepareTemplateProps } from '../../../lib/templating';
import defaultTemplates from '../../../widgets/clear-refinements/defaultTemplates';

describe('ClearRefinements', () => {
  const defaultProps = {
    refine: () => {},
    cssClasses: {
      root: 'root',
      button: 'button',
      disabledButton: 'disabled',
    },
    hasRefinements: true,
    templateProps: prepareTemplateProps({
      templates: {
        resetLabel: '',
      },
      defaultTemplates,
      templatesConfig: {},
    }),
    url: '#all-cleared!',
  };

  it('should render <ClearRefinements />', () => {
    const wrapper = mount(<ClearRefinements {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render <ClearRefinements /> with a specific class when no refinements', () => {
    const wrapper = mount(
      <ClearRefinements {...defaultProps} hasRefinements={false} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
