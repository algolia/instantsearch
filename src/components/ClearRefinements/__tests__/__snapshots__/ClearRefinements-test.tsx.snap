// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearRefinements should render <ClearRefinements /> 1`] = `
Array [
  <div
    className="root"
  >
    Array [
      <button
        className="button"
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
        disabled={false}
        onClick={[Function]}
      />,
    ]
  </div>,
]
`;

exports[`ClearRefinements should render <ClearRefinements /> with a specific class when no refinements 1`] = `
Array [
  <div
    className="root"
  >
    Array [
      <button
        className="button disabled"
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
        disabled={true}
        onClick={[Function]}
      />,
    ]
  </div>,
]
`;
